### DISCUSSION TITLE
Why should we choose  the replicatedmergeTree ?

### DISCUSSION THREAD

**Discussion by h1k** [ClickHouse/ClickHouse#48994] [Ideas]
When I use insert into distributed engine ,I happen to find that merge trees can synchronize data to replicas ,and without relying on zookeeper or Clickhouse keepers, so why is it recommended to replicate merge trees for replicas? What are the advantages of copying a merge tree over a merge tree?

My mergeTree example goes like this:
`CREATE TABLE default.test
(
id Int64
) ENGINE = Distributed('cluster_test', 'default', 'test_merge', rand())
`
`
CREATE TABLE default.test
(
id Int64
) ENGINE = MergeTree
ORDER BY tuple()
SETTINGS use_metadata_cache = 1
 `
My replicatedMergeTree example goes like this:
`CREATE TABLE default.test1
(
id Int64
) ENGINE = Distributed('cluster_test', 'default', 'test1_rep', rand())
`
`
CREATE TABLE default.test1_rep
(
id Int64
) Engine= ReplicatedMergeTree('/clickhouse/tables/test1_rep/{shard}', '{replica}') 
order by tuple() 
SETTINGS use_metadata_cache = 1
 `

**Comment by den-crane** [comment]
Because MergeTree does not guarantee consistency. Eventually MergeTree replicas will have different data.

Also there is huge maintenance issue how to add new replica / how to replace the lost / broken replica. Somehow you need manually copy data from the existing replica.

**Reply by h1k** [reply]
> Because MergeTree does not guarantee consistency. Eventually MergeTree replicas will have different data.

in my tests, the data of this copy is consistent, Is this accidental?How can I reproduce the scenario where copies are inconsistent?

**Reply by den-crane** [reply]
Because of network glitches and power outages. Distributed table re-tries insertion until success, if a network packet with success packet is lost, then Distributed table will repeat insertion and will create duplicate. Though nowadays it can be mitigated using `non_replicated_deduplication_window`. Also MergeTree tables are not durable, if power outage happens, MergeTree lost several recent inserts, but it's replica does not.

**Comment by lingtaolf** [comment]
IMO, here are several benefits to using ReplicatedMergeTree:
1. Data can be stored redundantly which means the maximum number of server unavailable is 1/2 of the number of server in your cluster;
2. Based 1, when you update your clickhouse version, you can update one by one so that the whole cluster can still provide service with correct result. ( In fact, different may have different data.)

**Reply by den-crane** [reply]
@lingtaolf you misunderstood the question. Clickhouse has 2 different ways to replicate data, using replicated engines (internal) and using distributed table. Distributed table is able to multiplex inserts into several replicas.
